File: src/engine/minhash.py
"""
Created on Mon Dec 17 13:44:23 2018

@author: julesbaudet
"""

import random
import time

maxArticleID = 2**32-1
# http://compoasso.free.fr/primelistweb/page/prime/liste_online_en.php
prime = 4294967311
# We generate random hash function of the form h_i(x) = (ax + b) mod(c)
#a and b are randomly generated coefficients, x is the number of the articles
#c has to be prime so that... COMPLETE

#first we create two lists of random coefficients
def RandomCoeffs(k):
    """generates a list of k different random coefficients"""
    t0 = time.time()
    coeffs = []
    while k>0:
        a = random.randint(0, maxArticleID)
        if a not in coeffs: 
            coeffs.append(a)
        k -= 1
    total_time = time.time() - t0
    print('It took {} time to create the coefficients for the hash functions \n'.format(total_time))
    return coeffs 

#we will only work on the articles that were cited by the users in our user list
def article_list(UserList):
    """ returns the list of all articles related to the users I'm considering"""
    t0 = time.time()
    articles = []
    for user in UserList:
        for article in user:
            if article not in articles:
                articles.append(article)
    total_time = time.time() - t0
    print('It took {} time to return the lists of articles considered \n'.format(total_time))
    return articles

#we now compute the signatures of all users
#we present it as a list "signature" such that signature[i] corresponds to user i's signature

def minhash(UserList, k):
    """ input a matrix with rows corresponding to the articles and column 
    corresponding to users, returns the matrix of signatures. k is the number of hash functions generated"""  
    t0 = time.time()
    
    coeff_a = RandomCoeffs(k)
    coeff_b = RandomCoeffs(k)
    all_signatures = []
    for user in UserList:
        #the signature of my user (all minhashes computed)
        user_signature = []
        for i in range(k):
            #we get the hash of each article
            #to be able to make comparisons, we initialize the variable min_Hash 
            #We give it a value that's bigger than the maximum value taken by our hash functions 
            min_hash = prime + 1
            for article in user:
                #compute the hash value
                hash_val = (coeff_a[i] * article + coeff_b[i]) % prime
                #update the value to get the minimum hash value obtained
                if hash_val < min_hash:
                    min_hash = hash_val
            user_signature.append(min_hash)
        all_signatures.append(user_signature)
    #the matrix all_signatures has rows corresponding to users and column corresponding to the hash function number
    total_time = time.time() - t0
    print(all_signatures[0:5][:50])
    print('It took {} time to create the signature matrix \n'.format(total_time))
    return all_signatures




def similarity(UserList, k):
    """returns the similarity matrix of users. We build it to be an upper triangular matrix not to waste storage or time"""
    t0 = time.time()
    
    signature_matrix = minhash(UserList, k)
    nb_usr = len(UserList)
    nb_hash = len(signature_matrix[0])
    #here we build a full matrix. Possibility: create a 1D array corresponding to a triangular matrix to optimize
    Sim_matrix = [[0 for _ in range(nb_usr)] for _ in range(nb_usr)]
    #take a user i
    for i in range(nb_usr):
        sig_i = signature_matrix[i]
        #we compare user i to all users after him (we build an upper triangular matrix: user i was already compared with users with smaller indexes)
        for j in range(i+1, nb_usr):
            sig_j = signature_matrix[j]
            
            #we now count the number of values in the signature that are equal
            count = 0
            for h in range(nb_hash):
                if sig_i[h] == sig_j[h]:
                    count += 1
            #compute the similarity percentage between user i and user j
            Sim_matrix[i][j] = count/nb_hash
    total_time = time.time() - t0
    print('It took {} time to to create the similarity matrix \n'.format(total_time))
    return Sim_matrix
        


#a little test program :)
def test_matrix(nb_usr = 500, nb_articles = 100):
    """generates a list of lists, where 1 list corresponds to a user and contains a list of articles"""
    #smallest ID of all articles
    t0 = time.time()
    min_articleID = 1000000000
    UserList = []
    for i in range(nb_usr):
        #number of articles cited by the User
        nb_cited = random.randint(1, 100)
        #list of articles cited by i
        art_i = []
        while nb_cited > 0:
            a = random.randint(min_articleID, min_articleID + nb_articles)
            if a not in art_i:
                nb_cited -= 1
                art_i.append(a)
        UserList.append(art_i)
    total_time = time.time() - t0
    print('It took {} time to to create the test matrix \n'.format(total_time))
    return UserList
        
def general_test(k):
    """tests the whole shit for k hash functions"""
    UserList = test_matrix()
    simi = similarity(UserList, k)
    print(simi[0][:100])
    print(simi[1][:100])
    print(UserList[0:3])

File: src/engine/MinHash.cpp
// Author: Jules BAUDET
#include <iostream>
#include <unordered_map>
#include <vector>

#include "MinHash.hpp"
#include <random>

int random_int() {
    // function that generates a random integer between 0 and 1500000
    int min = 0;
    int max = 1500000;
    std::random_device rd;  // used once to initialise engine
    std::mt19937 rng(rd()); // random-number engine
    std::uniform_int_distribution<int> uni(min, max);

    int random_integer = uni(rng);
    return random_integer;
}

MinHash::MinHash(
    int rank) { // creates random coefficients for the hash functions (rank = nb of hash functions)
    // our hash functions are of the form (a*articleID + b)%prime
    hash_nb = rank;
    int k = rank;
    while (k > 0) {
        int a = random_int();
        if (std::find(hashes_a.begin(), hashes_a.end(), a) == hashes_a.end()) {
            hashes_a.push_back(a); // we add a to our list of coefficients if it is not already in
                                   // it
            k = k - 1;
        };
    };

    int k2 = rank;
    while (k2 > 0) {
        int b = random_int(); // we repeat for the b coefficients
        if (std::find(hashes_b.begin(), hashes_b.end(), b) == hashes_b.end()) {
            hashes_b.push_back(b);
            k2 -= 1;
        };
    };
}

std::vector<int> MinHash::getSignature(const Author &author) {
    // returns the minhash signature of my author
    std::vector<int> user_signature;
    // for each hash function
    for (int i = 0; i < hash_nb; i++) {
        // initialize the minimum value to a value bigger than the maximum possibly obtained
        int minhash = prime + 1;
        // for each article cited by the author
        std::vector<std::string> paperVector = users2articles[author.name];
        for (int j = 0; j < paperVector.size(); j++) {
            // get the numerical ID of the article
            int article_nb = article2int[users2articles[author.name][j]];
            // hash this value (corresponds to a permutation)
            int hash_value = (article_nb * hashes_a[i] + hashes_b[i]) % prime;
            if (hash_value < minhash) {
                // update the value of minhash if the hash_value is lower than the current minhash
                minhash = hash_value;
            }
        }
        user_signature.push_back(minhash);
    }
    return user_signature;
}

float MinHash::getSimilarity(const Author &author1, const Author &author2) {
    // computes the similarity between 2 authors
    int count = 0;
    std::vector<int> signature1 = getSignature(author1);
    std::vector<int> signature2 = getSignature(author2);
    for (int i = 0; i < hash_nb; i++) {
        if (signature1[i] == signature2[i]) {
            count += 1; // we count the number of similarities in the signature
        }
    }
    return count; // the value of the similarity corresponds to cout/hash_nb
}

void MinHash::update(Edge edge) {
    // if the edge.author not in keys of users2articles, add it to users2articles
    // if the paper not in keys of articles2int, add it to article2int using the counter
    // add the paper to the vector at users2articles[user]

    // std::pair <std::unordered_map<std::string, std::vector<Paper> >::iterator, bool> tool;
    std::vector<std::string> vect;
    vect.push_back(edge.paper.id);
    // insert author in the map list (it will only do it if the author didn't exist before)
    auto tool = users2articles.insert(
        std::pair<std::string, std::vector<std::string>>(edge.author.name, vect));
    // if the author already is in the map keys (the author was seen before)
    if (tool.second == false) {
        if (std::find(users2articles[edge.author.name].begin(),
                      users2articles[edge.author.name].end(),
                      edge.paper.id) == users2articles[edge.author.name].end()) {
            users2articles[edge.author.name].push_back(edge.paper.id);
        }
    }
    // if the article was never seen before, we add it to our article2int map
    std::pair<std::unordered_map<std::string, int>::iterator, bool> ret2;
    ret2 = article2int.insert(std::pair<std::string, int>(edge.paper.id, counter));
    // if the articles was never seen before, it was added by the previous command
    if (ret2.second == true) {
        counter += 1;
    }
}

void MinHash::getInfo() {
    std::cout << "the hash number is " << hash_nb << std::endl;
    std::cout << "the number of authors is " << users2articles.size() << std::endl;
    std::cout << "the number of articles is " << counter << std::endl;
    std::cout << "counter = " << counter << std::endl;
    std::cout << "size = " << users2articles["1"].size() << std::endl;

    std::cout << "an example of authors article list is ";
    for (int i = 0; i < users2articles["1"].size(); i++) {
        std::cout << users2articles["1"].at(i) << ' ';
    }
    std::cout << " " << std::endl;

    std::cout << "an example of authors article list is ";
    for (int i = 0; i < users2articles["2"].size(); i++) {
        std::cout << users2articles["2"].at(i) << ' ';
    }
    std::cout << " " << std::endl;
}
